dm/syncer/binlog_locations.go

```diff
@@ -0,0 +1,175 @@
+	//   txnEndLocation will be assigned from curEndLocation
+	curStartLocation binlog.Location
+	curEndLocation   binlog.Location
+	txnEndLocation   binlog.Location
+
+	// DML will also generate a query event if user set session binlog_format='statement', we use this field to
+	// distinguish DML query event.
+	inDML bool
+
+	mu sync.Mutex // guard curEndLocation because Syncer.printStatus is reading it from another goroutine.
+}
+
+func (l *locationRecorder) reset(loc binlog.Location) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.curStartLocation = loc
+	l.curEndLocation = loc
+	l.txnEndLocation = loc
+}
+
+//nolint:unused
+func (l *locationRecorder) getCurEndLocation() binlog.Location {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.curEndLocation
+}
+
+//nolint:unused
+func (l *locationRecorder) setCurEndLocation(location binlog.Location) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.curEndLocation = location
+}
+
+// NOTE only locationRecorder should call this.
+func (l *locationRecorder) saveTxnEndLocation(needLock bool) {
+	if needLock {
+		l.mu.Lock()
+		defer l.mu.Unlock()
+	}
+
+	l.txnEndLocation = l.curEndLocation
+	_ = l.txnEndLocation.SetGTID(l.curEndLocation.GetGTID().Origin().Clone())
+}
+
+// shouldUpdatePos returns true when the given event is from a real upstream writing, returns false when the event is
+// header, heartbeat, etc.
+func shouldUpdatePos(e *replication.BinlogEvent) bool {
+	switch e.Header.EventType {
+	case replication.FORMAT_DESCRIPTION_EVENT, replication.HEARTBEAT_EVENT, replication.IGNORABLE_EVENT,
+		replication.PREVIOUS_GTIDS_EVENT, replication.MARIADB_GTID_LIST_EVENT:
+		return false
+	}
+	if e.Header.Flags&replication.LOG_EVENT_ARTIFICIAL_F != 0 {
+		// ignore events with LOG_EVENT_ARTIFICIAL_F flag(0x0020) set
+		// ref: https://dev.mysql.com/doc/internals/en/binlog-event-flag.html
+		return false
+	}
+
+	return true
+}
+
+func (l *locationRecorder) setCurrentGTID(gset mysql.GTIDSet) {
+	err := l.curEndLocation.SetGTID(gset)
+	if err != nil {
+		log.L().DPanic("failed to set GTID set",
+			zap.Any("GTID set", gset),
+			zap.Error(err))
+	}
+}
+
+// update maintains the member of locationRecorder as their definitions.
+// - curStartLocation is assigned to curEndLocation
+// - curEndLocation is tried to be updated in-place
+// - txnEndLocation is assigned to curEndLocation when `e` is the last event of a transaction.
+func (l *locationRecorder) update(e *replication.BinlogEvent) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.curStartLocation = l.curEndLocation
+
+	if !shouldUpdatePos(e) {
+		// so for these events, we'll set streamer to let curStartLocation equals curEndLocation
+		return
+	}
+
+	if event, ok := e.Event.(*replication.RotateEvent); ok {
+		nextName := string(event.NextLogName)
+		if l.curEndLocation.Position.Name != nextName {
+			l.curEndLocation.Position.Name = nextName
+			l.curEndLocation.Position.Pos = 4
+			l.saveTxnEndLocation(false)
```

Comment:
seems that all call of `saveTxnEndLocation` is `needLock=false`,`needLock = true` is it used later in the pr ?

Reply:
I'm not sure about future usage

Reply:
how about provide two functions instead of a boolean parameter